src/lib/proposer/proposer.ml

```diff
@@ -495,25 +465,25 @@ module Make (Inputs : Inputs_intf) :
                             Error.raise (Error.of_string str) )) )
         in
         let proposal_supervisor = Singleton_supervisor.create ~task:propose in
-        let keypairs_agent = Agent.create ~f:Fn.id keypairs_pipe in
-        let scheduler = Singleton_scheduler.create time_controller in
-        let last_keypairs =
-          Agent.get keypairs_agent |> Option.value ~default:Keypair.Set.empty
+        let last_keypairs_agent =
+          Agent.create ~f:Fn.id (Agent.Read_only.get keypairs)
         in
+        let scheduler = Singleton_scheduler.create time_controller in
         let rec check_for_proposal () =
           trace_recurring_task "check for proposal" (fun () ->
               (* See if we want to change keypairs *)
-              let keypairs =
-                Agent.get keypairs_agent
-                |> Option.value ~default:Keypair.Set.empty
-              in
-              if not (Keypair.Set.equal keypairs last_keypairs) then
+              let keypairs = Agent.Read_only.get keypairs in
+              let last_keypairs = Agent.get last_keypairs_agent in
+              if
+                not
+                  (Keypair.And_compressed_pk.Set.equal keypairs last_keypairs)
```

Comment:
Could the logic for updating consensus live in the API layer instead? This makes you pay `O(n)` every time you check for a proposal (which is often). Alternatively, we could track a hash of the set so that the comparison here is a scalar operation (and then the agent would calculate a hash whenever it gets written to). You can even make a simple abstraction on an agent that notifies you whether or not the value has changed since you last checked. Not too important now, but we should at least track an issue for it.

Reply:
Fixed via agent dirty bit